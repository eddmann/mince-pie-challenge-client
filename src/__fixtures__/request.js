export default function createRequestStack(responses = []) {
  const push = ({ path, method, body, response }) => responses.push({ path, method, body, response });

  const api = (path, method, body = undefined) => {
    const actual = { path, method, body };

    if (responses.length === 0) {
      return fail(`Expected: Nil, Actual: ${JSON.stringify(actual)}`);
    }

    const { response, ...expected } = responses.shift();

    expect(expected).toEqual(actual);

    return Promise.resolve(response);
  };

  const upload = (url, file) => api(url, 'PUT', file);

  return { push, api, upload };
};
