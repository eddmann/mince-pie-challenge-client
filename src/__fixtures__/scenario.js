export default function createScenario(fixtures = () => ({})) {
  return (label, run) => {
    const steps = [];

    run({
      given: (label, step) => steps.push(step),
      when: (label, step) => steps.push(step),
      then: (label, step) => steps.push(step),
      fixtures: fixtures(),
    });

    it(label, () =>
      steps.reduce((state, step) => state
        .then(wrapper => step(wrapper) || wrapper)
        .then(wrapper => new Promise(res => process.nextTick(() => { wrapper && wrapper.update(); res(wrapper); }))),
      Promise.resolve(undefined)));
  };
};
