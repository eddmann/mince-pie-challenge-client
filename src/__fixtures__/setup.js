import React from 'react';
import Enzyme, { mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import createMemoryHistory from 'history/createBrowserHistory';
import createScenario from './scenario';
import createRequestStack from './request';
import app from '../app';

Enzyme.configure({ adapter: new Adapter() });

export const scenario = createScenario(() => {
  const history = createMemoryHistory();
  const request = createRequestStack();
  const auth = { getToken: () => Promise.resolve() };

  return {
    request,
    auth,
    renderApp: (path = '/') => {
      history.replace(path);
      return app({
        history,
        services: { request, auth },
        links: { add: 'add.url', list: 'list.url', view: 'view.url/{id}' },
      }).then(mount);
    },
  };
});
