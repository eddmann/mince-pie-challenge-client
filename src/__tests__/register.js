import { scenario } from '../__fixtures__/setup';

scenario('registers a new user', ({ given, when, then, fixtures: { auth, renderApp } }) => {
  const USERNAME = 'sample_username';
  const EMAIL = 'sample@email.com';
  const PASSWORD = 'sample_password';
  const CODE = '1234';

  given('an unauthenicated user on the register page', () => {
    auth.register = jest.fn(() => Promise.resolve(USERNAME));
    auth.confirmRegistration = jest.fn(() => Promise.resolve());

    return renderApp('/register');
  });

  when('register with valid details', wrapper => {
    wrapper.find('input#username').simulate('change', { target: { id: 'username', value: USERNAME }});
    wrapper.find('input#email').simulate('change', { target: { id: 'email', value: EMAIL }});
    wrapper.find('input#password').simulate('change', { target: { id: 'password', value: PASSWORD }});
    wrapper.find('Button').simulate('click');
  });

  when('submit correct validation code', wrapper => {
    wrapper.find('input#code').simulate('change', { target: { id: 'code', value: CODE }});
    wrapper.find('Button').simulate('click');
  });

  then('successfully register and be redirected to login page', wrapper => {
    expect(auth.register).toBeCalledWith(USERNAME, EMAIL, PASSWORD);
    expect(auth.confirmRegistration).toBeCalledWith(USERNAME, CODE);
    expect(wrapper.find('.Login')).toMatchSnapshot();
  });
});
