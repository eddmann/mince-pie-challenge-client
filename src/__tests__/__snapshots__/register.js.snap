// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`registers a new user 1`] = `
<div
  className="Login"
>
  <Grid
    textAlign="center"
  >
    <div
      className="ui center aligned grid"
    >
      <GridColumn
        className="Form"
      >
        <div
          className="column Form"
        >
          <Header
            as="h2"
            color="teal"
            textAlign="center"
          >
            <h2
              className="ui teal center aligned header"
            >
              Login
            </h2>
          </Header>
          <Form
            as="form"
            size="large"
          >
            <form
              className="ui large form"
              onSubmit={[Function]}
            >
              <Segment
                stacked={true}
              >
                <div
                  className="ui stacked segment"
                >
                  <FormInput
                    as={[Function]}
                    control={[Function]}
                    fluid={true}
                    icon="user"
                    iconPosition="left"
                    id="username"
                    onChange={[Function]}
                    placeholder="Username"
                    value=""
                  >
                    <FormField
                      control={[Function]}
                      fluid={true}
                      icon="user"
                      iconPosition="left"
                      id="username"
                      onChange={[Function]}
                      placeholder="Username"
                      value=""
                    >
                      <div
                        className="field"
                      >
                        <Input
                          fluid={true}
                          icon="user"
                          iconPosition="left"
                          id="username"
                          onChange={[Function]}
                          placeholder="Username"
                          type="text"
                          value=""
                        >
                          <div
                            className="ui fluid left icon input"
                          >
                            <input
                              id="username"
                              key="text"
                              onChange={[Function]}
                              placeholder="Username"
                              type="text"
                              value=""
                            />
                            <Icon
                              as="i"
                              key="user"
                              name="user"
                            >
                              <i
                                aria-hidden="true"
                                className="user icon"
                              />
                            </Icon>
                          </div>
                        </Input>
                      </div>
                    </FormField>
                  </FormInput>
                  <FormInput
                    as={[Function]}
                    control={[Function]}
                    fluid={true}
                    icon="lock"
                    iconPosition="left"
                    id="password"
                    onChange={[Function]}
                    placeholder="Password"
                    type="password"
                    value=""
                  >
                    <FormField
                      control={[Function]}
                      fluid={true}
                      icon="lock"
                      iconPosition="left"
                      id="password"
                      onChange={[Function]}
                      placeholder="Password"
                      type="password"
                      value=""
                    >
                      <div
                        className="field"
                      >
                        <Input
                          fluid={true}
                          icon="lock"
                          iconPosition="left"
                          id="password"
                          onChange={[Function]}
                          placeholder="Password"
                          type="password"
                          value=""
                        >
                          <div
                            className="ui fluid left icon input"
                          >
                            <input
                              id="password"
                              key="password"
                              onChange={[Function]}
                              placeholder="Password"
                              type="password"
                              value=""
                            />
                            <Icon
                              as="i"
                              key="lock"
                              name="lock"
                            >
                              <i
                                aria-hidden="true"
                                className="lock icon"
                              />
                            </Icon>
                          </div>
                        </Input>
                      </div>
                    </FormField>
                  </FormInput>
                  <Button
                    as="button"
                    color="blue"
                    fluid={true}
                    loading={false}
                    onClick={[Function]}
                    size="large"
                  >
                    <button
                      className="ui blue large fluid button"
                      onClick={[Function]}
                      role="button"
                    >
                      Login
                    </button>
                  </Button>
                </div>
              </Segment>
            </form>
          </Form>
        </div>
      </GridColumn>
    </div>
  </Grid>
</div>
`;
